File: src/main/java/org/blaque/meetify/entity/Participant.java
package org.blaque.meetify.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "participants")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Participant {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String username;

    @Column(nullable = false, unique = true)
    private String sessionId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "room_id", nullable = false)
    private Room room;

    @Column(nullable = false)
    private LocalDateTime joinedAt;

    @Column
    private LocalDateTime leftAt;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isConnected = true;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isMuted = false;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isVideoEnabled = true;

    @PrePersist
    protected void onCreate() {
        joinedAt = LocalDateTime.now();
    }
}

File: frontend/src/domain/entities/Participant.ts
export class Participant {
  constructor(
    public readonly id: string,
    public readonly username: string,
    public readonly sessionId: string,
    public readonly joinedAt: Date,
    public readonly isConnected: boolean,
    public readonly isMuted: boolean,
    public readonly isVideoEnabled: boolean
  ) {}

  static fromDTO(dto: any): Participant {
    return new Participant(
      dto.id,
      dto.username,
      dto.sessionId,
      new Date(dto.joinedAt),
      dto.isConnected,
      dto.isMuted,
      dto.isVideoEnabled
    );
  }

  withConnectionStatus(isConnected: boolean): Participant {
    return new Participant(
      this.id,
      this.username,
      this.sessionId,
      this.joinedAt,
      isConnected,
      this.isMuted,
      this.isVideoEnabled
    );
  }

  withAudioStatus(isMuted: boolean): Participant {
    return new Participant(
      this.id,
      this.username,
      this.sessionId,
      this.joinedAt,
      this.isConnected,
      isMuted,
      this.isVideoEnabled
    );
  }

  withVideoStatus(isVideoEnabled: boolean): Participant {
    return new Participant(
      this.id,
      this.username,
      this.sessionId,
      this.joinedAt,
      this.isConnected,
      this.isMuted,
      isVideoEnabled
    );
  }
}
